src/api/order-v7.ts

```diff
@@ -0,0 +1,130 @@
+// save state
+import type { VercelRequest, VercelResponse } from '@vercel/node'
+import { MongoClient } from 'mongodb'
+import { decodeJWT } from '../auth'
+import { fulfillmentService, inventoryService, paymentService } from '../services'
+
+const mongo = new MongoClient(`mongodb://localhost`)
+
+async function retry<Result>(
+  serviceCall: () => Promise<Result>,
+  maxAttempts = 10,
+  callTimeout = 30 * 1000,
+  initialInterval = 1000
+): Promise<Result> {
+  let result: Result | undefined
+  let error: any
+  let attempt = 0
+  while (attempt < maxAttempts) {
+    try {
+      result = await Promise.race([
+        serviceCall(),
+        new Promise<undefined>((resolve) => setTimeout(resolve, callTimeout, undefined)),
+      ])
+      const timedOut = result === undefined
+      if (timedOut) {
+        throw new Error('Timed out')
+      }
+      break
+    } catch (e) {
+      error = e
+      await new Promise((resolve) => setTimeout(resolve, initialInterval * Math.pow(2, attempt++)))
+    }
+  }
+  if (error) {
+    throw error
+  }
+  if (!result) {
+    throw new Error('Reached max attempts')
+  }
+  return result
+}
+
+enum STATES {
+  CREATED,
+  RESERVED,
+  FAILED_TO_RESERVE,
+  PAID,
+  FAILED_TO_CHARGE,
+  FAILED_TO_CHARGE_UNRESERVED,
+  FULFILLED,
+  FAILED_TO_FULFILL,
+  FAILED_TO_FULFILL_UNRESERVED,
+  FAILED_TO_FULFILL_REFUNDED,
+}
+
+export default async (request: VercelRequest, response: VercelResponse) => {
+  const { itemId, quantity, addressId, requestId } = request.body
+  const { userId } = decodeJWT(request.headers.authorization)
+
+  try {
+    await mongo.connect()
+    const db = mongo.db('orders')
+    const orders = db.collection('orders')
+    const result = await orders.insertOne({
+      itemId,
+      quantity,
+      addressId,
+      requestId,
+      userId,
+      state: STATES.CREATED,
+    })
+    if (!result.acknowledged) {
+      response.status(500).send(`Unable to write to database`)
```

Comment:
Don't reveal there are database issues

```suggestion
      response.status(500).send('Failed to initiate order')
```